x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportStartTrainedModelDeploymentAction.java

```diff
@@ -509,6 +509,15 @@ public boolean test(ClusterState clusterState) {
                 return true;
             }
 
+            // We cannot add more nodes and the assignment is not satisfied
+            if (maxLazyMLNodes <= nodes.size()
+                && trainedModelAssignment.isSatisfied(nodes.stream().map(DiscoveryNode::getId).collect(Collectors.toSet())) == false) {
+                String msg = "Could not start deployment because there are not enough resources to provide all requested allocations";
+                logger.debug(() -> format("[%s] %s", modelId, msg));
+                exception = new ElasticsearchStatusException(msg, RestStatus.TOO_MANY_REQUESTS);
+                return true;
+            }
```

Comment:
We should also check to make sure that the current node size is less than the max node size (vertical scaling vs horizontal).

Reply:
Well spotted. I've added a commit that factors out a `isScalingPossible` and uses it in both places we make that check.